nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/StorageClientFactory.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.processors.azure.storage.utils;
+
+import com.azure.core.credential.AccessToken;
+import com.azure.core.credential.TokenCredential;
+import com.azure.core.http.HttpClient;
+import com.azure.core.http.ProxyOptions;
+import com.azure.core.http.netty.NettyAsyncHttpClientBuilder;
+import com.azure.identity.ClientSecretCredential;
+import com.azure.identity.ClientSecretCredentialBuilder;
+import com.azure.identity.ManagedIdentityCredential;
+import com.azure.identity.ManagedIdentityCredentialBuilder;
+import com.azure.storage.common.StorageSharedKeyCredential;
+import com.azure.storage.file.datalake.DataLakeServiceClient;
+import com.azure.storage.file.datalake.DataLakeServiceClientBuilder;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.nifi.services.azure.storage.ADLSCredentialsDetails;
+import reactor.core.publisher.Mono;
+
+public class StorageClientFactory {
+
+    private static final long STORAGE_CLIENT_CACHE_SIZE = 10;
+
+    private final Cache<Integer, DataLakeServiceClient> storageClientCache;
+
+    public StorageClientFactory() {
+        this.storageClientCache = createCache();
+    }
+
+    private Cache<Integer, DataLakeServiceClient> createCache() {
+        return Caffeine.newBuilder()
+                .maximumSize(STORAGE_CLIENT_CACHE_SIZE)
+                .build();
+    }
+
+    public Cache<Integer, DataLakeServiceClient> getCache() {
+        return storageClientCache;
+    }
+
+    /**
+     * Retrieves a {@link DataLakeServiceClient}
+     *
+     * @param proxyOptions not used for caching, because proxy parameters are set in the ProxyConfiguration service
+     * @param credentialsDetails used for caching because it can contain properties that are results of an expression
+     * @return DataLakeServiceClient
+     */
+    public DataLakeServiceClient getStorageClient(ProxyOptions proxyOptions, ADLSCredentialsDetails credentialsDetails) {
```

Comment:
The method can return the same `DataLakeServiceClient` object for multiple processors / processor threads.
Are we sure if `DataLakeServiceClient` is thread safe and can be used concurrently?

Reply:
I've checked the related code in `azure-storage-file-datalake-12.7.4.jar` and I didn't find anything that would made me think that the library is thread unsafe.

Reply:
@jfrazee Do you have any information whether `com.azure.storage.file.datalake.DataLakeServiceClient` is thread-safe or not?

Currently, `DataLakeServiceClient` is instantiated/built in every `onTrigger()` which leads to some performance loss. It is more relevant when a proxy is used (added in [NIFI-9951](https://issues.apache.org/jira/browse/NIFI-9951) recently), because the new client establishes a new connection to the proxy every time.
This PR is about caching the `DataLakeServiceClient` between `onTrigger()` calls but it is unclear if that client object can be used in parallel or not.
For the .NET ADLS library the is a clear statement that it is thread-safe:
https://docs.microsoft.com/en-us/dotnet/api/overview/azure/storage.files.datalake-readme
However, we cannot find a similar docs regarding the Java lib.

Could you please check it? Thanks in advance!

Reply:
I need to see whether this extends to the proxy scenario, but in the basic case, for recent versions of the SDK, the *ServiceClient can be treated as threadsafe:

https://github.com/Azure/azure-storage-java/blob/master/V12%20Upgrade%20Story.md

> v12 [...] Some of the highlights include: [...] Thread-safety and immutability. It is still safe to share clients across threads, and there should still be no confusion about when the client is in sync with the service as there is no state maintained on the client. State may be returned as the result of a method call, but there is no pretense of storing state on a client that could be immediately invalid.

https://techcommunity.microsoft.com/t5/azure-storage-blog/announcing-the-azure-storage-v12-client-libraries/ba-p/1482394

> All of our v12 libraries have improved performance and ensure thread safety.

https://azure.microsoft.com/en-us/blog/build-richer-applications-with-the-new-asynchronous-azure-storage-sdk-for-java/

> Azure Storage SDK v10 for Java adopts the next-generation Storage SDK design providing thread-safe types [...] Some of the improvements in the new SDK are: [...] Thread-safe interfaces[.]

https://azure.github.io/azure-sdk/java_introduction.html#service-client

>  DO ensure that all service client classes are immutable and stateless upon instantiation.

Reply:
We wasn't even sure that the basic scenario is thread safe. Thank you for the clarification! I think we can assume that the proxy scenario is thread safe as well.

Reply:
@jfrazee Do you have any update on the proxy scenario?
In my opinion, if a library/class declares itself as thread-safe (as the Azure v12 storage client in the referenced docs above), it must be thread-safe in all setup/usage. So I would go forward and merge this PR. Would you have any objections?